### ISSUE TITLE
Provide optional progress bar for intensive, incremental operations

### ISSUE THREAD

**Issue by sadielbartholomew** [NCAS-CMS/cf-python#183]

**Comment by davidhassell** [comment]
Hi @sadielbartholomew. Thank you for raising this issue.

I'm not really up on how the progress bars are calculated, but the dask progress bar appears to be the progress of the number of tasks (in the graph) completed by the scheduler. As far as I can gather, tqdm only keeps track of progress through top-level iterable.

Perhaps a double approach could be good: dask's ProgressBar for computations (when we have migrated to dask!) and tqdm for cf-python iterables (as found in cf.aggregate, for example).

What do you think?

**Comment by sadielbartholomew** [comment]
Thanks @davidhassell. You make very good points and generally they reminded me that the parallelism introduced by Dask will mean that progress isn't as simple to deduce as simple sequential iterations or similar. Overall I think this Issue should definitely be addressed only after #182, not just because this is bells and whistles' work rather than core work but because the under-the-hood logic behind the operations will change and the whole concept of progress with it.

> I'm not really up on how the progress bars are calculated, but the dask progress bar appears to be the progress of the number of tasks (in the graph) completed by the scheduler. As far as I can gather, tqdm only keeps track of progress through top-level iterable.

Indeed, it may be that we need a progress bar that is aware of the Dask task graph and means of computation etc. I'll bear it in mind whilst working on #182 but I suspect you are right that the Dask progress bar itself might be the only possibility.

> Perhaps a double approach could be good: dask's ProgressBar for computations (when we have migrated to dask!) and tqdm for cf-python iterables (as found in cf.aggregate, for example). What do you think?

That sounds like a good tentative plan. Ideally for simplicity and to not add another dependence we could stick to just one type, so it seems like the Dask one, but it might not be possible if there is a complicated mix of `cf`-only logic and logic deferred over Dask. Let's see what seems sensible once the LAMA to Dask migration is winding up.
